core/src/executor/datafusion/datafusion_processor.rs

```diff
@@ -144,18 +143,21 @@ impl DatafusionProcessor {
         // Conditionally create a new physical_plan if repartitioning is needed
         let plan_to_execute: Arc<dyn ExecutionPlan + 'static> =
             if physical_plan.output_partitioning().partition_count()
-                != self.config.target_partitions
+                != self.config.output_parallelism
             {
                 Arc::new(RepartitionExec::try_new(
                     physical_plan,
-                    Partitioning::RoundRobinBatch(self.config.target_partitions),
+                    Partitioning::RoundRobinBatch(self.config.output_parallelism),
                 )?)
             } else {
                 physical_plan
             };
 
-        let batches = execute_stream_partitioned(plan_to_execute, self.ctx.task_ctx())?;
-
+        // we must execute the plan with output parallelism, target_partitions is not used for repartitioning
+        let mut batches = Vec::with_capacity(self.config.output_parallelism);
```

Comment:
![image](https://github.com/user-attachments/assets/6fe3ffdd-a7ab-4ac7-86a2-ade91f928b8d)

Base on the definition of `RepartitionExec`, we don't need to manually repartition again. Do you have any test examples?